docs: overhaul contributing guide (#160)
confidential-containers/confidential-containers: CONTRIBUTING.md

and offer tips about how to work with the community and make great
contributions.

First off, Confidential Containers (CoCo) is an open source project.
You probably already knew that, but it's good to be clear
that we welcome contributions, involvement, and insight from everyone.
Just make sure to follow the [code of conduct](CODE_OF_CONDUCT.md).

CoCo is licensed with [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0).

This guide won't cover the architecture of the project, but you'll notice
that there are many different repositories under the [CoCo
organization](https://github.com/confidential-containers).
A CoCo deployment integrates many different components.
Fortunately the process for contributing to each of these subprojects
is largely the same.
There is one major exception. Some CoCo deployments leverage [Kata Containers](https://github.com/kata-containers),
which is an existing open source project. Kata has its own [Contributor Guide](https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md)
that you should take a look at.

## Connecting with the Community

You might already know exactly what you want to contribute and how.

> If not, if could be useful to interact with existing developers.

Reviewer: ```suggestion
If not, it would be useful to interact with existing developers.
```

There are several easy ways to do this.

### Slack Channel

The best place to interact with the CoCo community is the
`#confidential-containers` channel in the [CNCF Slack workspace](https://slack.cncf.io/).
This is a great place to ask any questions about the project
and to float ideas for future development.

### Community Meeting

CoCo also has a weekly community meeting on Thursdays.
See [the agenda](https://docs.google.com/document/d/1E3GLCzNgrcigUlgWAZYlgqNTdVwiMwCRTJ0QnJhLZGA/)
for more information.
The community meeting usually has a packed agenda, but there is always time
for a few basic questions.

### GitHub Issue

You can also open issues in GitHub.
Try to open your issue on the repository that is most closely related
to your question.
If you aren't sure which repository is most relevant, you can open an issue
on this repository.
If you're creating an issue that you plan to resolve, consider noting this in
a comment so other developers know that someone is working on the problem.

> ## Making a Contribution

Reviewer: ```suggestion
## Making Contributions
```

The mechanics of making a contribution are relatively straightforward.
All contributions should be made as GitHub pull requests.

> Make sure your PR includes a Developer Certificate of Origin

Reviewer: ```suggestion
Make sure your PR includes a Developer Certificate of Origin (DCO)
```

a few simple contributions to get used to the process.
Larger contributions, especially ones that span multiple subprojects,
or have implications on the trust model or project architecture,
will usually require more discussion and review.
Nonetheless, CoCo has a track record of [responding to PRs quickly](https://confidentialcontainers.devstats.cncf.io/d/10/pr-time-to-engagement)
and [merging PRs quickly](https://confidentialcontainers.devstats.cncf.io/d/16/opened-to-merged).

If you are preparing a large contribution, it is wise to share your idea
with the community as early as possible. Consider making an `RFC` issue
that explains the changes. You might also try to break large contributions
into smaller steps.

### Making a Pull Request

If you aren't familiar with Git or the GitHub PR workflow, take a look at [this section](https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md#github-workflow)
of the Kata Containers contributor guide.
We have a few firm formatting requirements that you must follow,
but in general if you are thoughtful about your work and responsive to review,
you shouldn't have any issues getting your PRs merged.

> The requirements that we do have are there to make everyone's life a little easier.

Reviewer 2: So Kata has requirements like the backport labels and references to tooling I don't think we've imported to coco. Would it be better to keep the github-workflow section in this document and cut out the kata-specific stuff?

Author: Let me change the wording a bit. Contributing to Kata is still a way to contribute to CoCo, so I think we should still point people to the Kata guide.

Don't worry if you forget to follow one of them at first.
You can always update your PR if necessary.

#### Certificate of Origin

> Every PR in every subproject is required to include a Developer Certificate of Origin (DCO).

Reviewer: ```suggestion
Every PR in every subproject is required to include a DCO.
```

This is strictly enforced by the CI.
Fortunately, it's easy to comply with this requirement.
At the end of the commit message for each of your commits add something like
```
Signed-off-by: Alex Ample <al@example.com>
```
You can add additional tags to credit other developers who worked on a commit
or helped with it.
See [here](https://ltsi.linuxfoundation.org/software/signed-off-process/)
for more information.

> #### Coding style

Reviewer 3: ```suggestion
#### Coding Style
```
For consistency

Please follow whatever stylistic and organizational conventions have been
established in the subproject that you are contributing to.

Most CoCo subprojects use Rust. When using Rust, it is a good idea to run `rustfmt` and `clippy`
before submitting a PR.
Most subprojects will automatically run these tools via a workflow,
but you can also run the tools locally.
In most cases your code will not be accepted if it does not pass these tests.

> Projects might have addtional conventions that are not captured by these tools.

Reviewer 2: ```suggestion
Projects might have additional conventions that are not captured by these tools.
```

> Projects might have addtional conventions that are not captured by these tools.

Reviewer: Can't we point to the `CONTRIBUTING.md` of each project individually? Taking this analogy further should we enlist each project individually here?

Author: I don't see contributing guides in many (any?) of the subprojects.

# Contributor Guide

So you want to contribute to Confidential Containers?
This guide will get you up to speed on the mechanics of the project
and offer tips about how to work with the community and make great
contributions.

First off, Confidential Containers (CoCo) is an open source project.
You probably already knew that, but it's good to be clear
that we welcome contributions, involvement, and insight from everyone.
Just make sure to follow the [code of conduct](CODE_OF_CONDUCT.md).

CoCo is licensed with [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0).

This guide won't cover the architecture of the project, but you'll notice
that there are many different repositories under the [CoCo
organization](https://github.com/confidential-containers).
A CoCo deployment integrates many different components.
Fortunately the process for contributing to each of these subprojects
is largely the same.

> There is one major exception. Some CoCo deployments leverage [Kata Containers](https://github.com/kata-containers),

Reviewer 4: Which ones do not leverage Kata?

Author: I changed this to "most deployments" since enclave-cc is the only one that does not.

## Connecting with the Community

You might already know exactly what you want to contribute and how.
If not, if would be useful to interact with existing developers.
There are several easy ways to do this.

### Slack Channel

The best place to interact with the CoCo community is the
`#confidential-containers` channel in the [CNCF Slack workspace](https://slack.cncf.io/).
This is a great place to ask any questions about the project
and to float ideas for future development.

### Community Meeting

CoCo also has a weekly community meeting on Thursdays.
See [the agenda](https://docs.google.com/document/d/1E3GLCzNgrcigUlgWAZYlgqNTdVwiMwCRTJ0QnJhLZGA/)
for more information.
The community meeting usually has a packed agenda, but there is always time
for a few basic questions.

### GitHub Issue

You can also open issues in GitHub.
Try to open your issue on the repository that is most closely related
to your question.
If you aren't sure which repository is most relevant, you can open an issue
on this repository.
If you're creating an issue that you plan to resolve, consider noting this in
a comment so other developers know that someone is working on the problem.

## Making Contributions

> The mechanics of making a contribution are relatively straightforward.

Reviewer 4: ```suggestion
The mechanics of making a contribution are straightforward and follow a typical [GitHub contribution flow](https://guides.github.com/introduction/flow/).
```

$ rustup component add rustfmt clippy
```

#### Commit Format

> Besides the code itself, your contribution will include commit messages,

Reviewer: I think this looks like one blob of text, having bullets will make it easier to follow?

Author: I shortened it to just include commit messages since the PR title and description are addressed in the next section.

word.

If the commit touches many different subsystems, you might want to
split it into multiple commits.
In general more smaller commits are preferred to fewer larger ones.

If a commit fixes an existing GitHub issue, include `Fixes: #xxx` in the commit message.
This will help GitHub link the commit to the issue, which shows other developers that
work is in progress.

#### Pull Request Format

The pull request itself must also have a name and a description.
Like a commit, the pull request name should start with a subsystem.
Unlike a commit, the pull request description does not become part of
the Git log.
The PR description can be more verbose. You might include a checklist of development steps
or questions that you would like reviewers to consider.
You can tag particular users or link a PR to an issue or another PR.
If your PR depends on other PRs, you should specify this in the description.
In general, the more context you provide in the description, the easier it will be for reviewers.

### Reviews

In most subprojects your PR must be approved by two reviewers before it is merged.
Reviews are also welcome from non-reviewers.

> For information about community structure, including how to become a reviewer,

Reviewer 2: The governance doc uses `maintainer` rather than `reviewer`. Consider adding a reviewer term in the governance or using maintainer here.

please see the [governance document](https://github.com/confidential-containers/confidential-containers/blob/main/governance.md).

It may take a few iterations to address concerns raised by maintainers.
Keep an eye on GitHub to see if you need to rebase your PR during this process.
The web UI will report whether a PR can be automatically merged or if it conflicts
with the base branch.

While CoCo reviewers tend to be responsive, many have a lot on their plate.
If your PR is awaiting review, you can tag a reviewer to remind them to take a look.
You can also ask for review in the CoCo Slack channel.

### Continuous Integration

All subprojects have some form of CI and require some checks to
pass before a PR can be merged.
Some of the checks are:

- Static analysis checks.
- Unit tests.
- Functional tests.
- Integration tests.

If your PR does not pass a check, try to find the cause of the failure.
If the cause is not related to your changes, you can ask a reviewer to re-run

> the tests or help troubleshoot.

Reviewer: ```suggestion
the tests or help troubleshoot.

With this guide, you're well-prepared to contribute effectively to Confidential Containers. We appreciate your involvement in our open-source community!
```
